### PR TITLE
Update Decent CI to use Python 3.12

### PR DESCRIPTION
**Pull request by Myoldmopar**
Python 3.8 hits EOL this year, and we are going to be building against 3.12 now so that we have several years of life without having to change again.  I am going to update the Decent CI runners to include 3.12.2 and update the Decent CI configs here accordingly.

### CODE REVIEW

**Review by jmarrec — COMMENTED** [review]
3 more platforms to go, but I like where this is going

**Review by jmarrec — APPROVED** [review]

File: .decent_ci-Linux.yaml [hunk 1]
```diff
@@ -1,14 +1,14 @@
 compilers:
   - name: "gcc"
     version: "11.4"
-    cmake_extra_flags: -DLINK_WITH_PYTHON:BOOL=ON -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA:STRING=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DBUILD_PERFORMANCE_TESTS:BOOL=ON -DVALGRIND_ANALYZE_PERFORMANCE_TESTS:BOOL=ON -DENABLE_PCH:BOOL=OFF
+    cmake_extra_flags: -DLINK_WITH_PYTHON:BOOL=ON -DPython_REQUIRED_VERSION:STRING=3.12 -DPython_ROOT_DIR:PATH=~/.pyenv/versions/3.12.2/ -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA:STRING=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DBUILD_PERFORMANCE_TESTS:BOOL=ON -DVALGRIND_ANALYZE_PERFORMANCE_TESTS:BOOL=ON -DENABLE_PCH:BOOL=OFF
```
Comment (jmarrec) [inline .decent_ci-Linux.yaml]: This is all fine, and explicit. I just wonder why you didn't set 3.12.2 as Python_REQUIRED_VERSION

**Review by jmarrec — APPROVED** [review]

File: .decent_ci-Linux.yaml (hunk 1 above)
Comment (Myoldmopar) [reply .decent_ci-Linux.yaml]: I just checked the commit history and found where I had "undone" this change before.  And it only had the X.Y version number, so mimicked that.  But I went ahead and put 3.12.2 on there now.

File: .decent_ci-Linux.yaml [hunk 2]
```diff
@@ -1,14 +1,14 @@
 compilers:
   - name: "gcc"
     version: "11.4"
-    cmake_extra_flags: -DLINK_WITH_PYTHON:BOOL=ON -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA:STRING=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DBUILD_PERFORMANCE_TESTS:BOOL=ON -DVALGRIND_ANALYZE_PERFORMANCE_TESTS:BOOL=ON -DENABLE_PCH:BOOL=OFF
+    cmake_extra_flags: -DLINK_WITH_PYTHON:BOOL=ON -DPython_REQUIRED_VERSION:STRING=3.12.2 -DPython_ROOT_DIR:PATH=~/.pyenv/versions/3.12.2/ -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA:STRING=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DBUILD_PERFORMANCE_TESTS:BOOL=ON -DVALGRIND_ANALYZE_PERFORMANCE_TESTS:BOOL=ON -DENABLE_PCH:BOOL=OFF
```
Comment (Myoldmopar) [inline .decent_ci-Linux.yaml:4]: OK, explicitly require 3.12.2 and also point it to the pyenv installed version.

File: .decent_ci-MacOS.yaml [hunk 3]
```diff
@@ -1,3 +1,3 @@
 compilers:
   - name: clang
-    cmake_extra_flags: -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DBUILD_FORTRAN=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON
+    cmake_extra_flags: -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DBUILD_FORTRAN=ON -DBUILD_TESTING:BOOL=ON -DENABLE_REGRESSION_TESTING:BOOL=ON -DREGRESSION_BASELINE_PATH:PATH=$REGRESSION_BASELINE -DREGRESSION_SCRIPT_PATH:PATH=$REGRESSION_DIR -DREGRESSION_BASELINE_SHA:STRING=$REGRESSION_BASELINE_SHA -DCOMMIT_SHA=$COMMIT_SHA -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON -DPython_REQUIRED_VERSION:STRING=3.12.2
```
Comment (Myoldmopar) [inline .decent_ci-MacOS.yaml:3]: On Mac, I'm not as sure.  IT helped me run the official Python 3.12.2 installer, so the /usr/bin/python3 is now 3.12.2.  I'm not _sure_ we need this addition, but it should be fine in this explicit declaration.  Do we need anything else specified?

File: .decent_ci-Windows.yaml [hunk 4]
```diff
@@ -2,5 +2,5 @@ compilers:
   - name: Visual Studio
     version: 16
     architecture: Win64
-    cmake_extra_flags: -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DCOMMIT_SHA=%COMMIT_SHA% -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON -DPython_EXECUTABLE:PATH=C:/Users/elee/AppData/Local/Programs/Python/Python311/python.exe
+    cmake_extra_flags: -DBUILD_FORTRAN:BOOL=ON -DBUILD_TESTING:BOOL=ON -DCOMMIT_SHA=%COMMIT_SHA% -DENABLE_GTEST_DEBUG_MODE:BOOL=OFF -DLINK_WITH_PYTHON=ON -DPython_EXECUTABLE:PATH=C:/Users/elee/AppData/Local/Programs/Python/Python312/python.exe
```
Comment (Myoldmopar) [inline .decent_ci-Windows.yaml:5]: Easy bump on Windows.

**Review by Myoldmopar — COMMENTED** [review]
I made sure all the Python installs also get `pip install boto beautifulsoup4 soupsieve`, so if CI runs on this latest commit and is happy, I think we're probably good.

**Comment by Myoldmopar** [comment]
> 3 more platforms to go, but I like where this is going

Windows, Mac, and Linux now done (pending CI results).  Is there a fourth?

File: .decent_ci-Windows.yaml (hunk 4 above)
Comment (jmarrec) [reply .decent_ci-Windows.yaml:5]: This one uses python_executable not python_root_dir
Which is fine but differs

**Comment by jmarrec** [comment]
> > 3 more platforms to go, but I like where this is going
> 
> Windows, Mac, and Linux now done (pending CI results).  Is there a fourth?

What? 1 + 3 = 3! 

I thought there was a specific yaml for the linux with coverage runner. I haven't looked at decent ci in a long time :)

**Comment by Myoldmopar** [comment]
Hey hey all green.  Merging this.  FYI @jmarrec !

**Pull request merged**